glib/src/main_context_futures.rs

```diff
@@ -230,19 +246,191 @@ impl TaskSource {
 
                 // This will panic if the future was a local future and is called from
                 // a different thread than where it was created.
-                Pin::new(&mut self.future).poll(&mut context)
+                if let Some(tx) = self.return_tx.take() {
```

Comment:
IIRC tokio for example still catches panics in these cases but just ignores them. For consistency we might want to do the same here

Reply:
The reason this is like this is so `MainContext::block_on` just propagates the panic up the stack.  Side effect is that `spawn_obj` and `spawn_local_obj` also do that so they kill the process on panic. Should they ignore panics? I am not sure the tokio behavior.

Reply:
Seems OK either way so let's just keep it like this I guess?

> The reason this is like this is so MainContext::block_on can pass None to just propagate the panic up the stack. 

As the C functions are not `extern "C-unwind"`, this will not actually happen in practice (or at least in the future)

Reply:
@jf2048 So what do you think? Let's get it in like this or do you want to change something? It seems fine to me either way for now

Reply:
This is just how it was before, IMO anything else to deal with the FFI problem should be part of #74

Reply:
Yeah I agree, I'm just asking because you seem to have explicitly implemented it the way it is because you thought that panics would be propagated out of `block_on()`

Reply:
Currently they are, even though it technically is undefined behavior

Reply:
I should clarify, it should be like this anyway even if the behavior is to abort on panic. Without this, a panic inside `block_on` is quietly ignored.

To really come up with a good fix for it, we would have to find some way to propagate the panic from any source func to the `g_main_loop_run` call. And of course in all other places where we pass callbacks to a C function.

Reply:
Good point. I don't like the quiet ignoring but OTOH that's what tokio does, and also what we would do here if people just ignore the join handle. Maybe if the channel for the join handle is closed, we should not catch the panic either but rethrow it?

> And of course in all other places where we pass callbacks to a C function.

I think this will only be possible case by case. In some cases we could simply mark the C callbacks as `C-unwind` instead of `C`, but it needs careful checking if this wouldn't introduce memory leaks or worse.

Reply:
> ...  but rethrow it?

Also for the other cases btw, not just the futures case

Reply:
@jf2048 What do you think? :)

Reply:
I suppose what we should do is always catch the panic and log it with `g_critical` if the channel is dropped from either end. There is no way to rethrow it in that case. It is not reliable to fall back on the default rust behavior right now since we are always in an FFI stack frame.

The other options are to make it into an abort, or just wait to see if abort becomes the default for `extern "C"`...

Reply:
`g_critical()` would be the tokio approach, right? If so, let's go with that

Reply:
Actually I don't think anything needs to be done, I was under the impression tokio set the hook but it does not. It just uses the default hook to print the message and ignores the panic: [current_thread.rs#L481-L483](https://github.com/tokio-rs/tokio/blob/39766220f4adc969ef7d026e04d752f5dfe55fb9/tokio/src/runtime/scheduler/current_thread.rs#L481-L483).

It also has an unstable feature that lets you configure the behavior: [https://docs.rs/tokio/latest/tokio/runtime/struct.Builder.html#method.unhandled_panic](Builder::unhandled_panic). But it "may change or be removed in the future". Practically for users it seems nearly the same as just setting `panic=abort`.

Reply:
I made a couple changes, it should now match the default behavior of `tokio`:

- `MainContext::spawn` and `MainContext::spawn_local` will propagate the panic into the `JoinHandle` and ignore it if dropped
- `MainContext::block_on` will rethrow the panic
- `Spawn` and `LocalSpawn` will just ignore the panic. Based on the behavior in `futures_executor`, it seems with those the caller is expected to use [`FuturesExt::catch_unwind`](https://docs.rs/futures/latest/futures/future/trait.FutureExt.html#method.catch_unwind) anyway if they want to change the behavior.

That should be good for now, if someone asks for some more ways to configure this we could add those in later